Build fail on gitlab CI job

Hello,

First, thank you for taking that project over - sphinx really needs proper multi-version support, outside of RTD.

Issue context:
We have a "docs" repository hosted on a GitLab instance, and we're having a CI job building the doc upon merge-requests (equivalent of the pull-request + github workflow).

Maybe I'm missing some options or something? Thank you for your help!

Versions of sphinx related packages:
```
$ pip freeze | grep sphinx
sphinx-autobuild==2024.2.4
sphinx-rtd-theme==2.0.0
sphinx-versioned-docs==1.2
sphinx_collapse==0.1.2
sphinxcontrib-applehelp==1.0.8
sphinxcontrib-devhelp==1.0.6
sphinxcontrib-htmlhelp==2.0.5
sphinxcontrib-jquery==4.1
sphinxcontrib-jsmath==1.0.1
sphinxcontrib-qthelp==1.0.7
sphinxcontrib-serializinghtml==1.1.10
sphinxemoji==0.3.1
```

The job is configured as follow:
```YAML
build-test:
  only:
    - merge_requests
  stage: ensure-build
  script: &script
    - make ci_build
  artifacts:
    expose_as: 'Docs preview'
    paths:
      - public/index.html
      - public
```

The `make ci_build` is as follow:
```
SPHINXBUILD       = sphinx-versioned
BUILDDIR              = ./public
ALLSPHINXOPTS  = --chdir sources -O $(BUILDDIR) --git-root ./ --local-conf ./sources/conf.py

.PHONY: ci_build
ci_build:
        $(SPHINXBUILD) $(ALLSPHINXOPTS)
        @echo
        @echo "Build finished. The HTML pages are in $(BUILDDIR)."
```

The issue looks like this (full output):
```
$ make ci_build
sphinx-versioned --chdir sources -O ./public --git-root ./ --local-conf ./sources/conf.py
| SUCCESS  | - latched into the git repo
| SUCCESS  | - located conf.py
| INFO     | - Pre-building...
╭───────────────────── Traceback (most recent call last) ──────────────────────╮
│ /usr/local/lib/python3.11/site-packages/sphinx_versioned/__main__.py:307 in  │
│ main                                                                         │
│                                                                              │
│   304 │                                                                      │
│   305 │   log.remove()                                                       │
│   306 │   log.add(sys.stderr, format=logger_format, level=loglevel.upper())  │
│ ❱ 307 │   return VersionedDocs(locals())                                     │
│   308                                                                        │
│   309                                                                        │
│   310 if __name__ == "__main__":                                             │
│                                                                              │
│ ╭──────────────────── locals ─────────────────────╮                          │
│ │                     chdir = 'sources'           │                          │
│ │                  git_root = './'                │                          │
│ │                local_conf = './sources/conf.py' │                          │
│ │                  loglevel = 'info'              │                          │
│ │               main_branch = 'main'              │                          │
│ │                output_dir = './public'          │                          │
│ │                  prebuild = True                │                          │
│ │                     quite = True                │                          │
│ │ reset_intersphinx_mapping = False               │                          │
│ │           select_branches = None                │                          │
│ │      sphinx_compatibility = False               │                          │
│ │                   verbose = False               │                          │
│ ╰─────────────────────────────────────────────────╯                          │
│                                                                              │
│ /usr/local/lib/python3.11/site-packages/sphinx_versioned/__main__.py:57 in   │
│ __init__                                                                     │
│                                                                              │
│    54 │   │   self._versions_to_pre_build = (                                │
│    55 │   │   │   self._select_specific_branches() if self.select_branches e │
│    56 │   │   )                                                              │
│ ❱  57 │   │   self._pre_build()                                              │
│    58 │   │                                                                  │
│    59 │   │   # Adds our extension to the sphinx-config                      │
│    60 │   │   application.Config = ConfigInject                              │
│                                                                              │
│ ╭───────────────────────────────── locals ─────────────────────────────────╮ │
│ │ config = {                                                               │ │
│ │          │   'chdir': 'sources',                                         │ │
│ │          │   'output_dir': './public',                                   │ │
│ │          │   'git_root': './',                                           │ │
│ │          │   'local_conf': './sources/conf.py',                          │ │
│ │          │   'reset_intersphinx_mapping': False,                         │ │
│ │          │   'sphinx_compatibility': False,                              │ │
│ │          │   'prebuild': True,                                           │ │
│ │          │   'select_branches': None,                                    │ │
│ │          │   'main_branch': 'main',                                      │ │
│ │          │   'quite': True,                                              │ │
│ │          │   ... +2                                                      │ │
│ │          }                                                               │ │
│ │   self = <sphinx_versioned.__main__.VersionedDocs object at              │ │
│ │          0x7f9630dfca90>                                                 │ │
│ ╰──────────────────────────────────────────────────────────────────────────╯ │
│                                                                              │
│ /usr/local/lib/python3.11/site-packages/sphinx_versioned/__main__.py:195 in  │
│ _pre_build                                                                   │
│                                                                              │
│   192 │   │   log.info("Pre-building...")                                    │
│   193 │   │                                                                  │
│   194 │   │   # get active branch                                            │
│ ❱ 195 │   │   self._active_branch = self.versions.active_branch              │
│   196 │   │                                                                  │
│   197 │   │   for tag in self._versions_to_pre_build:                        │
│   198 │   │   │   log.info(f"pre-building: {tag}")                           │
│                                                                              │
│ ╭───────────────────────────────── locals ─────────────────────────────────╮ │
│ │ self = <sphinx_versioned.__main__.VersionedDocs object at                │ │
│ │        0x7f9630dfca90>                                                   │ │
│ ╰──────────────────────────────────────────────────────────────────────────╯ │
│                                                                              │
│ /usr/local/lib/python3.11/site-packages/sphinx_versioned/versions.py:94 in   │
│ active_branch                                                                │
│                                                                              │
│    91 │   │   """Property to get active_branch."""                           │
│    92 │   │   if self._active_branch:                                        │
│    93 │   │   │   return self._active_branch                                 │
│ ❱  94 │   │   return self.repo.active_branch                                 │
│    95 │                                                                      │
│    96 │   pass                                                               │
│    97                                                                        │
│                                                                              │
│ ╭───────────────────────────────── locals ─────────────────────────────────╮ │
│ │   args = ()                                                              │ │
│ │ kwargs = {}                                                              │ │
│ │   self = <sphinx_versioned.versions.GitVersions object at                │ │
│ │          0x7f963071a390>                                                 │ │
│ ╰──────────────────────────────────────────────────────────────────────────╯ │
│                                                                              │
│ /usr/local/lib/python3.11/site-packages/git/repo/base.py:958 in              │
│ active_branch                                                                │
│                                                                              │
│    955 │   │   :return: Head to the active branch                            │
│    956 │   │   """                                                           │
│    957 │   │   # reveal_type(self.head.reference)  # => Reference            │
│ ❱  958 │   │   return self.head.reference                                    │
│    959 │                                                                     │
│    960 │   def blame_incremental(self, rev: str | HEAD, file: str, **kwargs: │
│    961 │   │   """Iterator for blame information for the given file at the g │
│                                                                              │
│ ╭─────────────────────────── locals ───────────────────────────╮             │
│ │ self = <git.repo.base.Repo '/builds/ci-framework/docs/.git'> │             │
│ ╰──────────────────────────────────────────────────────────────╯             │
│                                                                              │
│ /usr/local/lib/python3.11/site-packages/git/refs/symbolic.py:383 in          │
│ _get_reference                                                               │
│                                                                              │
│   380 │   │   │   to a reference, but to a commit"""                         │
│   381 │   │   sha, target_ref_path = self._get_ref_info(self.repo, self.path │
│   382 │   │   if target_ref_path is None:                                    │
│ ❱ 383 │   │   │   raise TypeError("%s is a detached symbolic reference as it │
│   384 │   │   return self.from_path(self.repo, target_ref_path)              │
│   385 │                                                                      │
│   386 │   def set_reference(                                                 │
│                                                                              │
│ ╭─────────────────────────── locals ───────────────────────────╮             │
│ │            self = <git.HEAD "HEAD">                          │             │
│ │             sha = '7a54f8d50349fee51789f0336634a27c6ebd5f21' │             │
│ │ target_ref_path = None                                       │             │
│ ╰──────────────────────────────────────────────────────────────╯             │
╰──────────────────────────────────────────────────────────────────────────────╯
TypeError: HEAD is a detached symbolic reference as it points to 
'7a54f8d50349fee51789f0336634a27c6ebd5f21'
make: *** [Makefile:18: ci_build] Error 1
Cleaning up project directory and file based variables 00:01
ERROR: Job failed: exit code 1
```

--- a/README.rst
+++ b/README.rst
@@ -35,10 +35,11 @@ How to use
     │ --sphinx-compatibility  -Sc                          Adds compatibility for older sphinx versions by monkey patching certain functions.                   │
     │ --prebuild                    --no-prebuild          Disables the pre-builds; halves the runtime [default: prebuild]                                      │
     │ --branches              -b                     TEXT  Build docs for specific branches and tags [default: None]                                            │
-    │ --main-branch           -m                     TEXT  Main branch to which the top-level `index.html` redirects to. [default: main]                        │
+    │ --main-branch           -m                     TEXT  Main branch to which the top-level `index.html` redirects to. Defaults to `main`. [default: None]    │
     │ --quite                       --no-quite             No output from `sphinx` [default: quite]                                                             │
     │ --verbose               -v                           Passed directly to sphinx. Specify more than once for more logging in sphinx.                        │
     │ --log                   -log                   TEXT  Provide logging level. Example --log debug, default=info [default: info]                             │
+    │ --force                                              Force branch selection. Use this option to build detached head/commits. [Default: False]             │
     │ --help                                               Show this message and exit.                                                                          │
     ╰───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
 

--- a/docs/changes/45.feature.rst
+++ b/docs/changes/45.feature.rst
@@ -1,2 +1,4 @@
 Adds the capability to build detached heads if either the head is already detached or that particular commit is
-specified via `--branches` arg, both of these will work if and only if `--force` is supplied too. 
+specified via `--branches` arg, provided that `--force` is supplied. Additionally, if the main-branch is not
+specified via `--main-branch` then the currently checkout out branch/tag will be considered as the main branch
+for generating the top-level `index.html`.

--- a/docs/changes/45.feature.rst
+++ b/docs/changes/45.feature.rst
@@ -1,0 +1,2 @@
+Adds the capability to build detached heads if either the head is already detached or that particular commit is
+specified via `--branches` arg, both of these will work if and only if `--force` is supplied too. 

--- a/sphinx_versioned/__main__.py
+++ b/sphinx_versioned/__main__.py
@@ -112,7 +112,7 @@ class VersionedDocs:
         _all_versions.extend(self.versions.repo.tags)
         _all_versions.extend(self.versions.repo.branches)
 
-        # check if the current git status is detached, if yes, append if `--force` is supplied
+        # check if `--force` is supplied, if yes, and if the current git status is detached, append:
         if self.force_branches:
             if self.versions.repo.head.is_detached:
                 log.warning(f"git detached {self.versions.repo.head.is_detached}")

--- a/sphinx_versioned/__main__.py
+++ b/sphinx_versioned/__main__.py
@@ -8,7 +8,7 @@ from sphinx import application
 from sphinx.errors import SphinxError
 from sphinx.cmd.build import build_main
 from sphinx.config import Config as SphinxConfig
-from sphinx_versioned.versions import GitVersions, BuiltVersions
+from sphinx_versioned.versions import GitVersions, BuiltVersions, PseudoBranch
 
 from loguru import logger as log
 
@@ -89,7 +89,7 @@ class VersionedDocs:
         `bool`
         """
         for tag in versions:
-            log.info(f"{msg}: {tag.name}")
+            log.debug(f"{msg}: {tag.name}")
         return True
 
     def _handle_paths(self) -> None:
@@ -111,23 +111,41 @@ class VersionedDocs:
         _all_versions = []
         _all_versions.extend(self.versions.repo.tags)
         _all_versions.extend(self.versions.repo.branches)
+
+        # check if the current git status is detached, if yes, append if `--force` is supplied
+        if self.force_branches:
+            if self.versions.repo.head.is_detached:
+                log.warning(f"git detached {self.versions.repo.head.is_detached}")
+                _all_versions.append(PseudoBranch(self.versions.repo.head.object.hexsha))
+
         self._log_versions(_all_versions)
         return _all_versions
 
     def _select_specific_branches(self) -> list:
-        _all_versions = self._get_all_versions()
+        log.debug(f"Instructions to select: `{self.select_branches}`")
+
+        _all_versions = {x.name: x for x in self._get_all_versions()}
         _select_specific_branches = []
 
-        for tag in _all_versions:
-            if tag.name in self.select_branches:
-                log.info(f"Selecting tag for further processing: {tag.name}")
-                _select_specific_branches.append(tag)
+        for tag in self.select_branches:
+            if tag in _all_versions.keys():
+                log.info(f"Selecting tag for further processing: {tag}")
+                _select_specific_branches.append(_all_versions[tag])
+            elif self.force_branches is True:
+                log.warning(f"Forcing build for branch `{tag}`, be careful, it may or may not exist!")
+                _select_specific_branches.append(PseudoBranch(tag))
+            else:
+                log.critical(f"Branch not found: `{tag}`, use `--force` to force the build")
+
         return _select_specific_branches
 
     def _generate_top_level_index(self) -> None:
         """Generate a top-level ``index.html`` with redirect to the main-branch version."""
         if self.main_branch not in [x.name for x in self._built_version]:
-            log.critical(f"main branch {self.main_branch} not found!!")
+            log.critical(
+                f"main branch `{self.main_branch}` not found!! / not building `{self.main_branch}`; top-level `index.html` will not be generated!"
+            )
+            return
 
         with open(self.output_dir / "index.html", "w") as findex:
             findex.write(
@@ -140,7 +158,6 @@ class VersionedDocs:
                 </head>
             """
             )
-        log.debug(f"main branch '{self.main_branch}' found")
         return
 
     def _build(self, tag, _prebuild=False) -> bool:
@@ -162,7 +179,7 @@ class VersionedDocs:
         EventHandlers.CURRENT_VERSION = tag
 
         with TempDir() as temp_dir:
-            log.debug(f"Checking out the latest tag in temporary directory: {temp_dir}")
+            log.debug(f"Checking out the tag in temporary directory: {temp_dir}")
             source = str(self.local_conf.parent)
             target = temp_dir
             argv = (source, target)
@@ -189,7 +206,7 @@ class VersionedDocs:
             self._versions_to_build = self._versions_to_pre_build
             return
 
-        log.info("Pre-building...")
+        log.debug("Pre-building...")
 
         # get active branch
         self._active_branch = self.versions.active_branch
@@ -283,6 +300,11 @@ def main(
     ),
     loglevel: str = typer.Option(
         "info", "-log", "--log", help="Provide logging level. Example --log debug, default=info"
+    ),
+    force_branches: bool = typer.Option(
+        False,
+        "--force",
+        help="Force branch selection. Use this option to build detached head/commits. [Default: False]",
     ),
 ) -> None:
     if select_branches:

--- a/sphinx_versioned/__main__.py
+++ b/sphinx_versioned/__main__.py
@@ -40,20 +40,22 @@ class VersionedDocs:
         self._parse_config(config)
         self._handle_paths()
 
-        self.output_dir = pathlib.Path(self.output_dir)
-
         self._versions_to_pre_build = []
         self._versions_to_build = []
         self._failed_build = []
-        self._additional_args = ()
-
-        self._additional_args += ("-Q",) if self.quite else ()
-        self._additional_args += ("-vv",) if self.verbose else ()
 
         # selectively build branches / build all branches
         self._versions_to_pre_build = (
             self._select_specific_branches() if self.select_branches else self._get_all_versions()
         )
+
+        # if `--force` is supplied with no `--main-branch`, make the `_active_branch` as the `main_branch`
+        if not self.main_branch:
+            if self.force_branches:
+                self.main_branch = self.versions.active_branch.name
+            else:
+                self.main_branch = "main"
+
         self._pre_build()
 
         # Adds our extension to the sphinx-config
@@ -70,6 +72,10 @@ class VersionedDocs:
     def _parse_config(self, config) -> bool:
         for varname, value in config.items():
             setattr(self, varname, value)
+
+        self._additional_args = ()
+        self._additional_args += ("-Q",) if self.quite else ()
+        self._additional_args += ("-vv",) if self.verbose else ()
         return True
 
     def _log_versions(
@@ -98,12 +104,16 @@ class VersionedDocs:
         log.debug(f"Working directory {self.chdir}")
 
         self.versions = GitVersions(self.git_root, self.output_dir)
+        self.output_dir = pathlib.Path(self.output_dir)
         self.local_conf = pathlib.Path(self.local_conf)
+
         if self.local_conf.name != "conf.py":
             self.local_conf = self.local_conf / "conf.py"
+
         if not self.local_conf.exists():
             log.error(f"conf.py does not exist at {self.local_conf}")
             raise FileNotFoundError(f"conf.py not found at {self.local_conf.parent}")
+
         log.success(f"located conf.py")
         return
 
@@ -131,7 +141,7 @@ class VersionedDocs:
             if tag in _all_versions.keys():
                 log.info(f"Selecting tag for further processing: {tag}")
                 _select_specific_branches.append(_all_versions[tag])
-            elif self.force_branches is True:
+            elif self.force_branches:
                 log.warning(f"Forcing build for branch `{tag}`, be careful, it may or may not exist!")
                 _select_specific_branches.append(PseudoBranch(tag))
             else:
@@ -147,6 +157,7 @@ class VersionedDocs:
             )
             return
 
+        log.success(f"main branch: `{self.main_branch}`; generating top-level `index.html`")
         with open(self.output_dir / "index.html", "w") as findex:
             findex.write(
                 f"""
@@ -286,10 +297,10 @@ def main(
         help="Build docs for specific branches and tags",
     ),
     main_branch: str = typer.Option(
-        "main",
+        None,
         "-m",
         "--main-branch",
-        help="Main branch to which the top-level `index.html` redirects to.",
+        help="Main branch to which the top-level `index.html` redirects to. Defaults to `main`.",
     ),
     quite: bool = typer.Option(True, help="No output from `sphinx`"),
     verbose: bool = typer.Option(

--- a/sphinx_versioned/_templates/versions.html
+++ b/sphinx_versioned/_templates/versions.html
@@ -98,25 +98,27 @@
 {% else %}
 <h3>{{ _('Versions') }}</h3>
 <ul>
-  {%- if versions.tags %}
-  <p>Tags</p>
-  <div class="dropdown-content">
-    {%- for name, url in versions.tags.items() %}
-    <div class="{{'rtd-current-item' if name==current_version }}">
-      <a href="{{ relpath }}{{ url }}">{{ name }}</a>
+  <div class="injected">
+    {%- if versions.tags %}
+    <p>Tags</p>
+    <div class="dropdown-content">
+      {%- for name, url in versions.tags.items() %}
+      <div class="{{'rtd-current-item' if name==current_version }}">
+        <a href="{{ relpath }}{{ url }}">{{ name }}</a>
+      </div>
+      {%- endfor %}
     </div>
-    {%- endfor %}
-  </div>
-  {%- endif %}
-  {%- if versions.branches %}
-  <p>Branches</p>
-  <div class="dropdown-content">
-    {%- for name, url in versions.branches.items() %}
-    <div class="{{'rtd-current-item' if name==current_version }}">
-      <a href="{{ relpath }}{{ url }}">{{ name }}</a>
+    {%- endif %}
+    {%- if versions.branches %}
+    <p>Branches</p>
+    <div class="dropdown-content">
+      {%- for name, url in versions.branches.items() %}
+      <div class="{{'rtd-current-item' if name==current_version }}">
+        <a href="{{ relpath }}{{ url }}">{{ name }}</a>
+      </div>
+      {%- endfor %}
     </div>
-    {%- endfor %}
+    {%- endif %}
   </div>
-  {%- endif %}
 </ul>
 {% endif %}

--- a/sphinx_versioned/versions.py
+++ b/sphinx_versioned/versions.py
@@ -1,5 +1,3 @@
-"""Collect and sort version strings."""
-
 import os
 import git
 import pathlib
@@ -7,6 +5,19 @@ from abc import ABC
 from loguru import logger as log
 
 os.environ["GIT_PYTHON_REFRESH"] = "quiet"
+
+
+class PseudoBranch:
+    """Class to generate a branchname for git detached situation"""
+
+    def __init__(self, name: str) -> None:
+        self.name = name
+        return
+
+    def __repr__(self) -> str:
+        return self.name
+
+    pass
 
 
 class _BranchTag(ABC):
@@ -91,6 +102,11 @@ class GitVersions(_BranchTag):
         """Property to get active_branch."""
         if self._active_branch:
             return self._active_branch
+
+        if self.repo.head.is_detached:
+            log.warning(f"git head detached: {self.repo.head.is_detached}")
+            return PseudoBranch(self.repo.head.object.hexsha)
+
         return self.repo.active_branch
 
     pass

--- a/sphinx_versioned/versions.py
+++ b/sphinx_versioned/versions.py
@@ -95,6 +95,7 @@ class GitVersions(_BranchTag):
             Name of branch/tag.
         """
         self._active_branch = name
+        log.debug(f"git checkout branch/tag: `{name}`")
         return self.repo.git.checkout(name, *args, **kwargs)
 
     @property

